x/tunnel/keeper/keeper_tunnel.go

```diff
@@ -0,0 +1,333 @@
+func (k Keeper) UpdateAndResetTunnel(
+	ctx sdk.Context,
+	tunnelID uint64,
+	signalDeviations []types.SignalDeviation,
+	interval uint64,
+) error {
+	tunnel, err := k.GetTunnel(ctx, tunnelID)
+	if err != nil {
+		return err
+	}
+
+	// edit the tunnel
+	tunnel.SignalDeviations = signalDeviations
+	tunnel.Interval = interval
+	k.SetTunnel(ctx, tunnel)
+
+	// edit the prices info
+	k.SetLatestPrices(ctx, types.NewLatestPrices(tunnelID, []feedstypes.Price{}, 0))
+
+	event := sdk.NewEvent(
+		types.EventTypeUpdateAndResetTunnel,
+		sdk.NewAttribute(types.AttributeKeyTunnelID, fmt.Sprintf("%d", tunnel.ID)),
+		sdk.NewAttribute(types.AttributeKeyInterval, fmt.Sprintf("%d", tunnel.Interval)),
+	)
+	for _, sd := range signalDeviations {
+		event = event.AppendAttributes(
+			sdk.NewAttribute(types.AttributeKeySignalID, sd.SignalID),
+			sdk.NewAttribute(types.AttributeKeySoftDeviationBPS, fmt.Sprintf("%d", sd.SoftDeviationBPS)),
+			sdk.NewAttribute(types.AttributeKeyHardDeviationBPS, fmt.Sprintf("%d", sd.HardDeviationBPS)),
+		)
+	}
+	ctx.EventManager().EmitEvent(event)
+
+	return nil
+}
+
+// SetTunnelCount sets the tunnel count in the store
+func (k Keeper) SetTunnelCount(ctx sdk.Context, count uint64) {
+	ctx.KVStore(k.storeKey).Set(types.TunnelCountStoreKey, sdk.Uint64ToBigEndian(count))
+}
+
+// GetTunnelCount returns the current number of all tunnels ever existed
+func (k Keeper) GetTunnelCount(ctx sdk.Context) uint64 {
+	return sdk.BigEndianToUint64(ctx.KVStore(k.storeKey).Get(types.TunnelCountStoreKey))
+}
+
+// SetTunnel sets a tunnel in the store
+func (k Keeper) SetTunnel(ctx sdk.Context, tunnel types.Tunnel) {
+	ctx.KVStore(k.storeKey).Set(types.TunnelStoreKey(tunnel.ID), k.cdc.MustMarshal(&tunnel))
+}
+
+// GetTunnel retrieves a tunnel by its ID
+func (k Keeper) GetTunnel(ctx sdk.Context, tunnelID uint64) (types.Tunnel, error) {
+	bz := ctx.KVStore(k.storeKey).Get(types.TunnelStoreKey(tunnelID))
+	if bz == nil {
+		return types.Tunnel{}, types.ErrTunnelNotFound.Wrapf("tunnelID: %d", tunnelID)
+	}
+
+	var tunnel types.Tunnel
+	k.cdc.MustUnmarshal(bz, &tunnel)
+	return tunnel, nil
+}
+
+// MustGetTunnel retrieves a tunnel by its ID. Panics if the tunnel does not exist.
+func (k Keeper) MustGetTunnel(ctx sdk.Context, tunnelID uint64) types.Tunnel {
+	tunnel, err := k.GetTunnel(ctx, tunnelID)
+	if err != nil {
+		panic(err)
+	}
+	return tunnel
+}
+
+// GetTunnels returns all tunnels
+func (k Keeper) GetTunnels(ctx sdk.Context) []types.Tunnel {
+	var tunnels []types.Tunnel
+	iterator := storetypes.KVStorePrefixIterator(ctx.KVStore(k.storeKey), types.TunnelStoreKeyPrefix)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var tunnel types.Tunnel
+		k.cdc.MustUnmarshal(iterator.Value(), &tunnel)
+		tunnels = append(tunnels, tunnel)
+	}
+	return tunnels
+}
+
+// ActiveTunnelID sets the active tunnel ID in the store
+func (k Keeper) ActiveTunnelID(ctx sdk.Context, tunnelID uint64) {
```

Comment:
Suggestion: Rename this function to `AddActiveTunnelID`

Reply:
Could we use this?
```
// SetActiveTunnelID sets the active tunnel ID in the store
func (k Keeper) SetActiveTunnelID(ctx sdk.Context, tunnelID uint64) {
	ctx.KVStore(k.storeKey).Set(types.ActiveTunnelIDStoreKey(tunnelID), []byte{0x01})
}

// DeleteActiveTunnelID deletes the active tunnel ID from the store
func (k Keeper) DeleteActiveTunnelID(ctx sdk.Context, tunnelID uint64) {
	ctx.KVStore(k.storeKey).Delete(types.ActiveTunnelIDStoreKey(tunnelID))
}
```